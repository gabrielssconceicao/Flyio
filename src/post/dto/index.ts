import { CreatePostDto } from './create-post.dto';
import { FindAllPostsResponseDto } from './find-all-posts.dto';
import { FindAllLikedPostsResponseDto } from './find-all-liked-post.dto';
export { CreatePostDto, FindAllPostsResponseDto, FindAllLikedPostsResponseDto };
