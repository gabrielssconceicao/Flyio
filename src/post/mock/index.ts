import { postMock } from './post.mock';
import { generateCreatePostDtoMock } from './create-post.dto.mock';
export { postMock, generateCreatePostDtoMock };
