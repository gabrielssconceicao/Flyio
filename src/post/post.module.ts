import { Module } from '@nestjs/common';
import { PostService } from './post.service';
import { PostController } from './post.controller';
import { PostCommentController } from './post-comment.controller';
import { PostCommentService } from './post-comment.service';

@Module({
  controllers: [PostController, PostCommentController],
  providers: [PostService, PostCommentService],
})
export class PostModule {}
