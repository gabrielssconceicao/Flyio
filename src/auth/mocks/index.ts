import { hashingServiceMock } from './hashing.service.mock';
import { jwtConfigurationMock } from './jwt.configuration.mock';
import { jwtServiceMock } from './jwt.service.mock';
import { generateLoginDtoMock } from './login.dto.mock';
import { generateTokenPayloadDtoMock } from './token-payload.dto.mock';

export {
  hashingServiceMock,
  jwtConfigurationMock,
  jwtServiceMock,
  generateLoginDtoMock,
  generateTokenPayloadDtoMock,
};
