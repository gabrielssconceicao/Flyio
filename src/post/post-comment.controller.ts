import {
  Body,
  Controller,
  HttpCode,
  HttpStatus,
  Param,
  Post,
  UseGuards,
} from '@nestjs/common';
import { ApiBearerAuth, ApiOperation } from '@nestjs/swagger';
import { PostCommentService } from './post-comment.service';
import { CreateCommentDto } from './dto/create-comment.dto';
import { TokenPayloadParam } from '../auth/params/token-payload.param';
import { TokenPayloadDto } from '../auth/dto';
import { ApiAuthResponses } from '../common/decorators/guard.decorator';
import { AuthTokenGuard } from '../auth/guard/auth-token.guard';

@Controller('post')
@ApiBearerAuth()
@ApiAuthResponses()
@UseGuards(AuthTokenGuard)
export class PostCommentController {
  constructor(private readonly postCommentService: PostCommentService) {}

  @ApiOperation({ summary: 'Create post comment' })
  @HttpCode(HttpStatus.CREATED)
  @Post(':postId/comment')
  create(
    @Param('postId') postId: string,
    @Body() body: CreateCommentDto,
    @TokenPayloadParam() tokenPayload: TokenPayloadDto,
  ) {
    return this.postCommentService.create(postId, body, tokenPayload);
  }
}
