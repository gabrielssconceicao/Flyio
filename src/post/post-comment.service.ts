import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { TokenPayloadDto } from '../auth/dto';
import { CreateCommentDto } from './dto/create-comment.dto';

@Injectable()
export class PostCommentService {
  constructor(private readonly prismaService: PrismaService) {}

  async create(
    postId: string,
    createCommentDto: CreateCommentDto,
    tokenPayload: TokenPayloadDto,
  ) {
    const post = await this.prismaService.post.findUnique({
      where: { id: postId },
    });
    if (!post) {
      throw new NotFoundException('Post not found');
    }
    const comment = await this.prismaService.comment.create({
      data: {
        content: createCommentDto.content,
        userId: tokenPayload.sub,
        postId,
      },
      select: {
        id: true,
        content: true,
        createdAt: true,
        user: {
          select: {
            name: true,
            username: true,
            profileImg: true,
          },
        },
      },
    });

    return comment;
  }
}
