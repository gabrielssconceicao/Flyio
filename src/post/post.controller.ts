import {
  Controller,
  Post,
  Body,
  Param,
  Delete,
  UseInterceptors,
  UploadedFiles,
  UseGuards,
  HttpStatus,
  Get,
  Query,
  HttpCode,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiBody,
  ApiConsumes,
  ApiOperation,
  ApiResponse,
} from '@nestjs/swagger';
import { FilesInterceptor } from '@nestjs/platform-express';
import * as multer from 'multer';
import { ApiAuthResponses } from '../common/decorators/guard.decorator';
import { PostService } from './post.service';
import { CreatePostDto } from './dto/create-post.dto';
import { AuthTokenGuard } from '../auth/guard/auth-token.guard';
import { PostImagesValidatorPipe } from '../cloudinary/pipes/post-image-validator.pipe';
import { TokenPayloadParam } from '../auth/params/token-payload.param';
import { TokenPayloadDto } from '../auth/dto';
import { PostEntity } from './entities/post.entity';
import { FindAllPostsResponseDto } from './dto/find-all-posts.dto';
import { PaginationDto } from '../common/dto/pagination.dto';

@Controller('post')
@ApiBearerAuth()
@ApiAuthResponses()
@UseGuards(AuthTokenGuard)
export class PostController {
  constructor(private readonly postService: PostService) {}

  @ApiOperation({ summary: 'Create a post for user' })
  @ApiConsumes('multipart/form-data')
  @ApiBody({
    description: 'Create a post for user',
    schema: {
      type: 'object',
      properties: {
        text: {
          type: 'string',
          description: 'Post text',
        },
        images: {
          type: 'array',
          items: {
            type: 'string',
            format: 'binary',
          },
          description: 'Image files (PNG or JPEG)',
        },
      },
      required: ['text'],
    },
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Post created successfully',
    type: PostEntity,
  })
  @ApiResponse({
    status: HttpStatus.BAD_REQUEST,
    description: 'File upload failed due to invalid file type or size limit.',
    schema: {
      example: {
        statusCode: HttpStatus.BAD_REQUEST,
        message: 'Invalid file type or file too small/too large',
        error: 'Bad Request',
      },
    },
  })
  @Post()
  @HttpCode(HttpStatus.CREATED)
  @UseInterceptors(
    FilesInterceptor('images', 4, {
      storage: multer.memoryStorage(),
    }),
  )
  create(
    @Body() createPostDto: CreatePostDto,
    @UploadedFiles(PostImagesValidatorPipe) images: Express.Multer.File[],
    @TokenPayloadParam() tokenPayload: TokenPayloadDto,
  ) {
    return this.postService.create(createPostDto, images, tokenPayload);
  }

  @ApiOperation({ summary: 'Get all posts' })
  @ApiResponse({
    status: HttpStatus.OK,
    description: 'Posts fetched successfully',
    type: FindAllPostsResponseDto,
  })
  @Get()
  @HttpCode(HttpStatus.OK)
  findAll(@Query() paginationDto: PaginationDto) {
    return this.postService.findAll(paginationDto);
  }

  @ApiOperation({ summary: 'Get post by id' })
  @ApiResponse({
    status: HttpStatus.OK,
    description: 'Post fetched successfully',
    type: PostEntity,
  })
  @ApiResponse({
    status: HttpStatus.NOT_FOUND,
    description: 'Post not found',
    schema: {
      example: {
        statusCode: HttpStatus.NOT_FOUND,
        message: 'Post not found',
        error: 'Not Found',
      },
    },
  })
  @HttpCode(HttpStatus.OK)
  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.postService.findOne(id);
  }

  @ApiOperation({ summary: 'Delete post by id' })
  @ApiResponse({
    status: HttpStatus.OK,
    description: 'Post deleted successfully',
  })
  @ApiResponse({
    status: HttpStatus.NOT_FOUND,
    description: 'Post not found',
    schema: {
      example: {
        statusCode: HttpStatus.NOT_FOUND,
        message: 'Post not found',
        error: 'Not Found',
      },
    },
  })
  @HttpCode(HttpStatus.OK)
  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.postService.remove(id);
  }
}
