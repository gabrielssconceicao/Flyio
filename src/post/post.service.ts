import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { CreatePostDto } from './dto/create-post.dto';
import { TokenPayloadDto } from '../auth/dto';
import { PrismaService } from '../prisma/prisma.service';
import { CloudinaryService } from '../cloudinary/cloudinary.service';
import { FindAllPostsResponseDto } from './dto/find-all-posts.dto';
import { PostEntity } from './entities/post.entity';
import { PaginationDto } from '../common/dto/pagination.dto';
import { PermissionService } from '../permission/permission.service';

@Injectable()
export class PostService {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly cloudinaryService: CloudinaryService,
    private readonly permissionService: PermissionService,
  ) {}

  private selectPostFields = {
    id: true,
    content: true,
    createdAt: true,
    user: {
      select: {
        username: true,
        profileImg: true,
      },
    },
    images: {
      select: {
        id: true,
        url: true,
      },
    },
    _count: {
      select: {
        PostLikes: true,
        comments: true,
      },
    },
  };

  async create(
    createPostDto: CreatePostDto,
    images: Express.Multer.File[],
    tokenPayload: TokenPayloadDto,
  ): Promise<PostEntity> {
    let postImagesUrl: string[] = [];
    if (images.length) {
      postImagesUrl = await this.cloudinaryService.uploadPostImages(images);
    }
    const { _count, ...post } = await this.prismaService.post.create({
      data: {
        content: createPostDto.text,
        userId: tokenPayload.sub,
        images: {
          createMany: {
            data: postImagesUrl.map((url) => ({ url })),
          },
        },
      },
      select: this.selectPostFields,
    });
    return { ...post, likes: _count.PostLikes, liked: false, commentsCount: 0 };
  }

  async findAll(
    paginationDto: PaginationDto,
    tokenPayload: TokenPayloadDto,
  ): Promise<FindAllPostsResponseDto> {
    const { limit = 50, offset = 0 } = paginationDto;
    const { count, items } = await this.prismaService.findAll(
      this.prismaService.post,
      {
        select: {
          ...this.selectPostFields,

          PostLikes: {
            where: { userId: tokenPayload.sub },
          },
        },

        take: limit,
        skip: offset,
        orderBy: {
          createdAt: 'desc',
        },
      },
    );

    return {
      count,
      items: items.map(({ _count, PostLikes, ...post }) => ({
        ...post,
        commentsCount: _count.comments,
        likes: _count.PostLikes,
        liked: PostLikes.length ? true : false,
      })),
    };
  }

  async findOne(id: string, tokenPayload: TokenPayloadDto) {
    const post = await this.prismaService.post.findUnique({
      where: { id },
      select: {
        ...this.selectPostFields,
        PostLikes: {
          where: { userId: tokenPayload.sub },
        },
        comments: {
          select: {
            id: true,
            content: true,
            createdAt: true,
            user: {
              select: {
                name: true,
                username: true,
                profileImg: true,
              },
            },
          },
          orderBy: {
            createdAt: 'desc',
          },
        },
      },
    });
    if (!post) {
      throw new NotFoundException('Post not found');
    }
    const { _count, PostLikes, comments, ...rest } = post;
    return {
      ...rest,
      likes: _count.PostLikes,
      commentsCount: _count.comments,
      comments,
      liked: !!PostLikes.length,
    };
  }

  async remove(id: string, tokenPayload: TokenPayloadDto): Promise<void> {
    const post = await this.findOne(id, tokenPayload);
    this.permissionService.verifyUserOwnership(
      post.user.username,
      tokenPayload.username,
    );
    if (post.images.length) {
      await this.cloudinaryService.deletePostImages(
        post.images.map((url) => url.url),
      );
    }

    await this.prismaService.post.delete({ where: { id } });
    return;
  }

  async like(postId: string, tokenPayload: TokenPayloadDto) {
    const post = await this.findOne(postId, tokenPayload);
    if (post.liked) {
      throw new BadRequestException('Post already liked');
    }
    await this.prismaService.postLikes.create({
      data: {
        postId,
        userId: tokenPayload.sub,
      },
    });
  }
  async unlike(postId: string, tokenPayload: TokenPayloadDto) {
    const post = await this.findOne(postId, tokenPayload);
    if (!post.liked) {
      throw new BadRequestException('Post not liked');
    }
    await this.prismaService.postLikes.delete({
      where: {
        postId_userId: {
          postId,
          userId: tokenPayload.sub,
        },
      },
    });
  }
}
