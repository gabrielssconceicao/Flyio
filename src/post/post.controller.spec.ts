import { Test, TestingModule } from '@nestjs/testing';
import { JwtService } from '@nestjs/jwt';
import { PostController } from './post.controller';
import { PostService } from './post.service';
import { jwtConfigurationMock, jwtServiceMock } from '../auth/mocks';
import jwtConfig from '../auth/config/jwt.config';
import {
  generateCreatePostDtoMock,
  generateFindAllPostsDtoMock,
  postMock,
} from './mock';
import { generateFileMock } from '../cloudinary/mocks';
import { generateTokenPayloadDtoMock } from '../auth/mocks';
describe('PostController', () => {
  let controller: PostController;
  let postService: PostService;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [PostController],
      providers: [
        {
          provide: PostService,
          useValue: {
            create: jest.fn(),
            findAll: jest.fn(),
          },
        },
        {
          provide: JwtService,
          useValue: jwtServiceMock,
        },
        {
          provide: jwtConfig.KEY,
          useValue: jwtConfigurationMock,
        },
      ],
    }).compile();

    controller = module.get<PostController>(PostController);
    postService = module.get<PostService>(PostService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
    expect(postService).toBeDefined();
  });

  describe('<Create />', () => {
    it('should create a post', async () => {
      jest.spyOn(postService, 'create').mockResolvedValue(postMock as any);
      const result = await controller.create(
        generateCreatePostDtoMock(),
        [generateFileMock()],
        generateTokenPayloadDtoMock(),
      );
      expect(result).toEqual(postMock);
      expect(postService.create).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });
  });

  describe('<FindAll/>', () => {
    it('should find all posts', async () => {
      jest
        .spyOn(postService, 'findAll')
        .mockResolvedValue(generateFindAllPostsDtoMock());

      const result = await controller.findAll({ limit: 10, offset: 0 });
      expect(postService.findAll).toHaveBeenCalled();
      expect(result).toEqual(generateFindAllPostsDtoMock());
    });
  });
});
