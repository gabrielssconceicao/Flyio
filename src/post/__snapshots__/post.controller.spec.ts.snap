// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostController <Create /> should create a post 1`] = `
{
  "commentsCount": 0,
  "content": "Hello world",
  "createdAt": 2023-03-15T17:30:00.000Z,
  "id": "42-d-f-df4",
  "images": [],
  "liked": false,
  "likes": 0,
  "user": {
    "profileImg": "https://example.com/profile.jpg",
    "username": "jDoe453",
  },
}
`;
